test/unit/reporters/html/html.test.js

```diff
@@ -123,3 +107,54 @@ describe('HTML Reporter', () => {
         });
     });
 });
+
+describe('HTML reporter --html-failed-only flag', () => {
+    beforeEach(() => {
+        sandbox.stub(view, 'save');
+        sandbox.stub(logger, 'log');
+        sandbox.stub(fs, 'copyAsync').returns(Promise.resolve());
+        sandbox.stub(fs, 'mkdirsAsync').returns(Promise.resolve());
+
+        emitter = new QEmitter();
+        emitter.config = {
+            forBrowser: sinon.stub().returns({
+                rootUrl: 'browser/root/url'
+            })
+        };
+    });
+
+    afterEach(() => sandbox.restore());
+
+    it('should copy reference screenshots when turned off', () => {
+        new HtmlReporter(emitter, null, {failedOnly: false}); // eslint-disable-line no-new
+        sandbox.stub(lib, 'referenceAbsolutePath').returns('absolute/reference/path');
+
+        emitter.emit(Events.TEST_RESULT, mkStubResult_({
+            referencePath: 'reference/path',
+            equal: true
+        }));
+
+        emitter.emit(Events.END);
+
+        return emitter.emitAndWait(Events.END_RUNNER).then(() => {
+            assert.calledOnce(fs.copyAsync);
+            assert.calledWith(fs.copyAsync, 'reference/path', 'absolute/reference/path');
+        });
+    });
+
+    it('should not copy reference screenshots when turned on', () => {
+        new HtmlReporter(emitter, null, {failedOnly: true}); // eslint-disable-line no-new
+        sandbox.stub(lib, 'referenceAbsolutePath').returns('absolute/reference/path');
+
+        emitter.emit(Events.TEST_RESULT, mkStubResult_({
+            referencePath: 'reference/path',
+            equal: true
+        }));
+
+        emitter.emit(Events.END);
+
+        return emitter.emitAndWait(Events.END_RUNNER).then(() => {
+            assert.notCalled(fs.copyAsync);
+        });
+    });
+});
```

Comment:
Tests updated, now there is one general describe with tests, which do not care about failedOnly option and then there are 2 describes inside for the failedOnly option. Hopefully, it looks better.